### PR TITLE
Fix: Add more space to bottom docked button

### PR DESCRIPTION
**Pull request by truph01**
### Fixed Issues
 
$ https://github.com/Expensify/App/issues/44056
PROPOSAL: https://github.com/Expensify/App/issues/44056#issuecomment-2179772696

### Tests

1. Open app
2. Tap avatar
4. Tap display name
5. Observe the space at the bottom of docked button
6. There should be proper bottom SafeSpace below the button
- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

Screenshot 2024-07-03 at 14 10 53

</details>

<details>
<summary>Android: mWeb Chrome</summary>

Screenshot 2024-07-03 at 14 22 15

</details>

<details>
<summary>iOS: Native</summary>

Screenshot 2024-07-03 at 14 09 59

</details>

<details>
<summary>iOS: mWeb Safari</summary>

Screenshot 2024-07-03 at 14 21 09

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

Screenshot 2024-07-03 at 14 19 09

</details>

<details>
<summary>MacOS: Desktop</summary>

Screenshot 2024-07-03 at 14 17 58

</details>

### CODE REVIEW

**Comment by shawnborton** [comment]
Looking good, but here are spots where it's missing:
- Wallet > Transfer Balance page
- Workspace settings > Members > Invite new members

On the Edit expense details > Date screen, we have too much space below the Save button:
CleanShot 2024-07-03 at 10 16 13@2x


One interesting observation though... I feel like when the keyboard is up, we don't need the SafeSpace to show. Is there something we can do about that? Otherwise it feels like the gap below the Save button is way too large when the keyboard is visible. 
CleanShot 2024-07-03 at 10 16 01@2x

**Comment by truph01** [comment]
@shawnborton 
> Looking good, but here are spots where it's missing:
Wallet > Transfer Balance page
Workspace settings > Members > Invite new members

- These screens do not use a form wrapper component, but the fix in this PR just applies to the bottom button in form wrapper component. If we need to fix these screen, we need to add the padding bottom to its bottom button manually. WDYT?

> On the Edit expense details > Date screen, we have too much space below the Save button:

- I fixed it

> One interesting observation though... I feel like when the keyboard is up, we don't need the SafeSpace to show. Is there something we can do about that? Otherwise it feels like the gap below the Save button is way too large when the keyboard is visible.

- I fixed it. Below is the result:

https://github.com/Expensify/App/assets/170341821/590d8171-4c52-41f0-8ee3-aeb183fb155f

**Comment by shawnborton** [comment]
> we need to add the padding bottom to its bottom button manually. WDYT?

I think we should add the padding there for consistency. And we should check as many other pages as we can too to make sure we aren't missing any. 

> I fixed it. Below is the result:

Much better!

**Comment by truph01** [comment]
@allgandalf I updated PR. Now, I use the additional padding bottom style in ```FormAlertWithSubmitButton``` directly instead of styling ```<FormAlertWithSubmitButton containerStyles={[..., additionalPaddingBottomStyle]} ... />``` in ```FormWrapper``` as I did before because, there is a lot of screens we do not use ```FormWrapper``` but still have docked button like mentioned above:

> Looking good, but here are spots where it's missing:
Wallet > Transfer Balance page
Workspace settings > Members > Invite new members

File: src/hooks/useExtraSafePaddingBottomStyle.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import useStyledSafeAreaInsets from './useStyledSafeAreaInsets';
+import useThemeStyles from './useThemeStyles';
+
+const useExtraSafePaddingBottomStyle = () => {
+    const styles = useThemeStyles();
+    const [willKeyboardShow, setWillKeyboardShow] = useState(false);
+    useEffect(() => {
+        const keyboardWillShowListener = Keyboard.addListener('keyboardWillShow', () => {
+            setWillKeyboardShow(true);
+        });
```
[the remark this answers is not in the document]
Comment (truph01) [reply src/hooks/useExtraSafePaddingBottomStyle.ts:12]: Ah, I missed your question. That ```useEffect``` will listen to the keyboard state to know when the keyboard will show or hide.

**Review by allgandalf — COMMENTED** [review]
also are we covering all the cases with these changes?

File: src/hooks/useExtraSafePaddingBottomStyle.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import {useEffect, useMemo, useState} from 'react';
+import {Keyboard} from 'react-native';
+import useStyledSafeAreaInsets from './useStyledSafeAreaInsets';
+import useThemeStyles from './useThemeStyles';
+
+const useExtraSafePaddingBottomStyle = () => {
+    const styles = useThemeStyles();
+    const [willKeyboardShow, setWillKeyboardShow] = useState(false);
```
[the remark this answers is not in the document]
Comment (truph01) [reply src/hooks/useExtraSafePaddingBottomStyle.ts:8]: - I don't think we should we the function you mentioned, because it is not used in our app, just in test file.
- We already have the way to get the keyboard state, see this [file](https://github.com/Expensify/App/blob/main/src/components/withKeyboardState.tsx). But it use ```keyboardDidShow``` and ```keyboardDidHide``` events to check if the keyboard is shown or not. It works well but there is an issue: when the keyboard is shown (user can see the keyboard), the bottom button is still applied the ```padding-bottom: 20px``` for a while. So I use ```keyboardWillShow``` and ```keyboardWillHide``` to fix that issue.

**Comment by truph01** [comment]
> also are we covering all the cases with these changes?

Yes

File: src/hooks/useExtraSafePaddingBottomStyle.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (truph01) [reply src/hooks/useExtraSafePaddingBottomStyle.ts]: I added the type for it. But I don't think we should rename the state, because as I mentioned above, ```willKeyboardShow``` is used to fix the issue ```didKeyboardShow``` encountered: There is a delay when the keyboard is visible by the user and the when ```didKeyboardShow``` is set to ```true```. So it is not just ```isKeyboardVisible```.

**Comment by truph01** [comment]
@allgandalf Plz help review PR again. Thanks

**Comment by shawnborton** [comment]
Is this one ready for testing again? Would love to get this one over the finish line!

**Comment by shawnborton** [comment]
Design-wise this is looking good, but let's still run it through a proper C+ testing. 

Can you try pulling main though so I can rebuild the test builds? From the test build above, a lot of the page transitions and such feel a bit jumpy or not very smooth... so just want to make sure this PR isn't causing any of that. Let me know if you are experiencing something similar.

**Comment by shawnborton** [comment]
Cool, will rebuild the test builds and see how it feels. I think we can move forward with the checklist and final review though in the meantime!

**Comment by shawnborton** [comment]
Okay nice, this feels way better. I say we get this into final review!

**Comment by shawnborton** [comment]
Hmm how are we calculating that space? How does the space below other buttons on this same platform look?

**Comment by shawnborton** [comment]
Yeah, that looks better. Can you help me understand why the two of them are different though?

**Review by allgandalf — APPROVED** [review]
changes LGTM, looks good to be merged into staging

**Comment by shawnborton** [comment]
Got it. Yeah, ideally we use whatever is consistent with other places (like display name). How do you recommend we fix that?

**Comment by shawnborton** [comment]
Sounds good!

File: src/hooks/useExtraSafePaddingBottomStyle.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import {useEffect, useMemo, useState} from 'react';
+import {Keyboard} from 'react-native';
+import useStyledSafeAreaInsets from './useStyledSafeAreaInsets';
+import useThemeStyles from './useThemeStyles';
+
+const useExtraSafePaddingBottomStyle = () => {
```
Comment (puneetlath) [inline src/hooks/useExtraSafePaddingBottomStyle.ts:6]: Perhaps we can name it something that makes it more clear when it can be used. "extra safe" is a bit ambiguous.

**Review by puneetlath — CHANGES_REQUESTED** [review]

**Comment by puneetlath** [comment]
@Expensify/design someone mind reviewing as well?

**Comment by puneetlath** [comment]
Oh whoops, my bad. I see @shawnborton already reviewed above.

**Comment by truph01** [comment]
@puneetlath I addressed your comment above

File: src/hooks/useSafePaddingBottomStyle.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import {useEffect, useMemo, useState} from 'react';
+import {Keyboard} from 'react-native';
+import useStyledSafeAreaInsets from './useStyledSafeAreaInsets';
+import useThemeStyles from './useThemeStyles';
+
+const useSafePaddingBottomStyle = () => {
```
Comment (puneetlath) [inline src/hooks/useSafePaddingBottomStyle.ts:6]: Can we add some general documentation above about what this hook is for.

File: src/components/FormAlertWithSubmitButton.tsx [hunk 5]
```diff
@@ -79,10 +80,11 @@ function FormAlertWithSubmitButton({
 }: FormAlertWithSubmitButtonProps) {
     const styles = useThemeStyles();
     const style = [!footerContent ? {} : styles.mb3, buttonStyles];
+    const extraSafePaddingBottomStyle = useSafePaddingBottomStyle();
```
Comment (puneetlath) [inline src/components/FormAlertWithSubmitButton.tsx:83]: To match the hook name. 
```suggestion
    const safePaddingBottomStyle = useSafePaddingBottomStyle();
```

**Review by puneetlath — CHANGES_REQUESTED** [review]
Sorry a couple more small requests.

**Comment by truph01** [comment]
All comment are resolved

File: src/hooks/useSafePaddingBottomStyle.ts (hunk 4 above)
Comment (puneetlath) [reply src/hooks/useSafePaddingBottomStyle.ts:9]: Sorry for being a stickler, but I'm finding this comment a bit hard to follow. In the future, when should someone use this hook? Maybe if you can just explain that to me in words in a comment here, I can help you craft a code comment that is a bit more clear.

**Review by puneetlath — CHANGES_REQUESTED** [review]

File: src/hooks/useSafePaddingBottomStyle.ts (hunk 4 above)
Comment (truph01) [reply src/hooks/useSafePaddingBottomStyle.ts:9]: @puneetlath

> In the future, when should someone use this hook

This hook, useSafePaddingBottomStyle, is useful for adding extra bottom padding to a component based on the device's safe area. If the device's safe area padding bottom is 0, the hook returns 0. Otherwise, it provides a padding bottom of 20. Use this to ensure content visibility and layout consistency across different devices.

Comment (puneetlath) [reply src/hooks/useSafePaddingBottomStyle.ts:9]: That's beautiful. Very clear! Let's make that the comment.

File: src/hooks/useSafePaddingBottomStyle.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
+import {useEffect, useMemo, useState} from 'react';
+import {Keyboard} from 'react-native';
+import useStyledSafeAreaInsets from './useStyledSafeAreaInsets';
+import useThemeStyles from './useThemeStyles';
+
+// Refer to https://github.com/Expensify/App/issues/44056, the default padding bottom of any device is not enough.
+// Devices with 0 padding bottom value, this hook returns { paddingBottom: 0 }.
+// Otherwise, it returns { paddingBottom: 20 }.
```
Comment (puneetlath) [inline src/hooks/useSafePaddingBottomStyle.ts:8]:
```suggestion
// This hook is useful for adding extra bottom padding to a component based on the device's safe area
// If the device's safe area padding bottom is 0, the hook returns 0. Otherwise, it provides a padding bottom of 20.
// Use this to ensure content visibility and layout consistency across different devices.
```

**Review by puneetlath — CHANGES_REQUESTED** [review]

**Comment by truph01** [comment]
I updated PR @puneetlath

**Review by puneetlath — APPROVED** [review]
